### ISSUE TITLE
Ignore parentless nodes

### ISSUE THREAD

**Issue by kevinruscoe** [BlueM/Tree#22]
I get my list from an external API, and it, unfortunately, contains parentless nodes. So when building a tree, it keeps throwing an exception "Node with ID $id points to non-existent parent with ID $pid".

What's the best approach for ignoring these?

**Comment by kevinruscoe** [comment]
Actually, if the "build()" method was public, I could just extend the class and overwrite that method.

**Comment by sippejw** [comment]
Ideally the InvalidParentException would contain information about the invalid node. So that a try catch could be utilized to remove the node and then recursively attempt to build the tree again.

**Comment by kevinruscoe** [comment]
Seems long-winded. I'm writing up a PR atm that basically has a "ignore" flag, which will silence the exceptions.

**Comment by BAGELreflex** [comment]
It's not that long winded. In an ideal world, the implementation could handle the missing record however it see's fit.

```php
class MyClass
{
    const ROOT_ID = -1;

    // ...

    public function dataToTree(array $data = [])
    {
        try
        {
            return new \BlueM\Tree(
                $data,
                ['rootId' => $this::ROOT_ID, 'id' => 'id', 'parent' => 'parent_id']
            );
        }
        catch (\BlueM\Tree\Exception\InvalidParentException $ex)
        {
            return $this->dataToTree(array_values(array_diff_keys($data, [$ex->errorNode->index =>''])));

            // This may be more efficient?
            // unset($data[$ex->errorNode->index]);
            // return $this->dataToTree(array_values($data));

            // Removing the item is about just as good as just ignoring it, but what if I wanted to load it?
            $data[] = $this->load($ex->errorNode->parent_id);
            return $this->dataToTree($data);
        }
    }
}
```

**Comment by kevinruscoe** [comment]
The exception is thrown each time a node is troublesome, so you could end up calling the `dataToTree()` method 100s of needless times until it eventually creates the correct tree. Simply ignoring these troublesome nodes doesn't cause any unforeseen problems (the OPs tests are still covered), they're just discarded from the tree. 

So, being able to silence exceptions, IMO, makes more sense.

**Comment by BAGELreflex** [comment]
I guess my argument really comes down to not having to parse the $ex->getMessage() using something like `` to know index caused the exception to be thrown. With that provided, actually catching the exception becomes much more useful.

```php
<?php
namespace BlueM\Tree\Exception;
/**
 * Exception which will be thrown if a tree node's parent ID points to an inexistent node.
 *
 * @author  Carsten Bluem <carsten@bluem.net>
 * @license http://www.opensource.org/licenses/bsd-license.php BSD 3-Clause License
 */
class InvalidParentException extends \RuntimeException
{
    private $_index;

    public function __construct(string $message = '', int $index = 0)
    {
        $this->_index = $index;
        parent::__construct($message);
    }

    public function getIndex()
    {
        return $this->_index;
    }
}
```

```php
// ...
foreach ($children as $pid => $childIds) {
	foreach ($childIds as $index => $id) {
		if ((string) $pid === (string) $id) {
			if (! $this->silenceBuildErrors) {
				throw new InvalidParentException(
					"Node with ID $id references its own ID as parent ID",
                                        $index // or $id
				);
			}
		}
		if (isset($this->nodes[$pid])) {
			$this->nodes[$pid]->addChild($this->nodes[$id]);
		} else {
			if (! $this->silenceBuildErrors) {
				throw new InvalidParentException(
					"Node with ID $id points to non-existent parent with ID $pid",
                                        $index // or $id
				);
			}
		}
	}
}
// ...
```

It would also be worth discussing if storing the index is as useful as storing `$id` (nodeId) in the above example. Storing `nodeId` would most likely result in the need to use `array_search()` to locate the record in the original data, so `index` would be the more useful data if the goal is to do something with the data. Storing `nodeId` would be much more useful for logging as well.

Silencing the exception allows me to achieve my actual desired outcome, but it leaves me with no visibility over if data was actually removed. If my resultset has thousands of items, it is very difficult to find what data may be missing. Storing just one additional piece of information greatly increases the usability of this class.

**Comment by BlueM** [comment]
Alternative approach: instead of `throw new InvalidParentException()` in case of an inconsistency, a new method `protected handleInconsistency()` (default implementation would simply run `throw new InvalidParentException()` to prevent the BC break) could be called with the error message – plus additional data, such as the node ID and parent ID – when an inconsistency is observed. 

That way, it was up to the implementation to decide what to do:
* Throw immediately
* Don’t do anything at all (= ignore)
* Don’t throw, but record the problems and offer a method to retrieve them.
* Do anything else which I can’t even imagine

This would also allow the method to differentiate between the problem cases (parent invalid vs. parent == self). Bonuses for code quality: it would not clutter `build()` with even more conditions and very good testability.

@BAGELreflex: reliable getting the problematic ID is surely useful. However, with the approach outlined above, it would simply no longer be necessary to parse anything, as `handleInconsistency()` would get that information as argument.

Questions:
1. What do you think?
2. Would you prefer a strategy approach, i.e. being able to inject some code instead of having the need for subclassing and overwriting `handleInconsistency()`?

**Comment by kevinruscoe** [comment]
Sounds good. Another idea I had late last night, would be to make a new BuilderInterface() and allow the user to define _how_ they want the tree. The default implementation would be yours. That way @BAGELreflex's implementation could throw more informative exceptions and my implementation would just ignore the problematic nodes.

**Comment by BAGELreflex** [comment]
Providing the ability to put my own callback handler would absolutely suffice. So, what do you suggest would be passed to `handleInconsitency()`? The entire current node that has an inconsistency? That seems the most useful and would reduce having to look it back up in the data if there was just an ID or index provided.

So, for @kevinruscoe to ignore his node, a basic callback could just be provided, like:
```php
$this->inconsistencyHandler = function($node) {};
```

This may actually end up requiring the use of `call_user_func()`.

**Comment by BlueM** [comment]
@kevinruscoe: I don’t agree to the `BuilderInterface`, for 2 reasons:

1. In the described use case (simple ignore errors, don't change anything else) you would have to overwrite the whole method just for the purpose of ignoring.
2. The rest of the code expects the data to be structured in a certain way – which is produced in `build()`, and therefore replacing `build()` could break things. (Although I don’t remember exactly, I suppose that’s why I made `build()` private)

Despite reason 2, I tend to change `private` to `protected`, as there might be valid use cases for overwriting.

Regardless of a protected `build()`, I still like the idea of a flexible inconcistency handling mechanism (either by subclassing or by passing in a `callable`). I will post a suggestion.

**Comment by BlueM** [comment]
@kevinruscoe: I just realized that in your first comment, you wrote “parentless nodes”. Which probably means there is not an invalid parent ID, but no property / array key for the parent ID at all? If yes, suppressing the exception would not have much effect, as you would still get an “Undefined index” warning in https://github.com/BlueM/Tree/blob/master/src/Tree.php#L222. So, for cases like these, a different approach must be taken.

**Comment by BAGELreflex** [comment]
@BlueM I think that would be an argument for a different issue. One that focuses on handling malformed data rather than missing data. Though, the implementation for that would be very similar to how this resolution will be implemented. I believe that too will require a defaulthandler which throws a new exception. This would also have to modify the build() process to do error checking/handling. I will open a new issue for this one with some options for handling it.

**Comment by jonagoldman** [comment]
Why is `null` not allowed as a root id?

Line 65:

```php
if (isset($options['rootid'])) { // this s
    if (!\is_scalar($options['rootid'])) {
        throw new \InvalidArgumentException('Option “rootid” must be a scalar');
    }
    $this->rootId = $options['rootid'];
}
```

1) Why is `rootid` lowercase (instead of `rootId`) ?
2) Why `isset($options['rootid'])` instead of `array_key_exists('rootid', $options)` ?
3) Why `!\is_scalar($options['rootid'])` ?

**Comment by BlueM** [comment]
To all who commented in this issue: I just pushed 604d6ea0e54e954aa4ef4f3e7c4b229c85001ae3 in a new branch “build-flexibility“, which adds a new constructor option “buildWarningCallback”, which can be set to a callable which will be called in case of a problem. The signature of the callable should be like that of method `buildWarningHandler()`, which is the default implementation and which behaves like previous versions, so the default behavior is unchanged. Using this approach, you can silence errors (by passing in an empty function, which does _not_ throw an exception), throw any exception you like or try to fix the node. Moreover, `build()` is now `protected`, which opens new ways for pre-checking or pre-modifying the data.

@jonagoldman: this is a legitimate question (although I would have preferred it as a separate issue). I don’t see why `null` should not be used – semantically, it probably makes even more sense than `0` –, so the branch also changes that.

**Comment by Pnoexz** [comment]
@BlueM any estimated date of release for the `build-flexibility` branch? I'm looking forward to using both functionalities it provides, but can't modify the minimum stability in my composer.json

**Comment by BlueM** [comment]
@Pnoexz: I think I’ll wait at least until the weekend to give the participants in this discussion time to react.

**Comment by BlueM** [comment]
The branch is now in `master` and included in version 3.1
